src/setinverse.jl

```diff
@@ -12,7 +12,9 @@ struct FunctionWithInverse{F,InvF} <: Function
     f::F
     invf::InvF
 end
-
+FunctionWithInverse(::Type{F}, invf::InvF) where {F,InvF} = FunctionWithInverse{Type{F},InvF}(F,invf)
+FunctionWithInverse(f::F, ::Type{InvF}) where {F,InvF} = FunctionWithInverse{F,Type{InvF}}(f,InvF)
+FunctionWithInverse(::Type{F}, ::Type{InvF}) where {F,InvF} = FunctionWithInverse{Type{F},Type{InvF}}(F,InvF)
```

Comment:
```suggestion
FunctionWithInverse(::Type{F}, invf::InvF) where {F,InvF} = FunctionWithInverse{F,InvF}(F,invf)
FunctionWithInverse(f::F, ::Type{InvF}) where {F,InvF} = FunctionWithInverse{F,InvF}(f,InvF)
FunctionWithInverse(::Type{F}, ::Type{InvF}) where {F,InvF} = FunctionWithInverse{F,InvF}(F,InvF)
```

`::Type` in a signature is only used so that you can dispatch on the exact identity of a type, as otherwise the type is `DataType`. For example:

```julia
julia> f(::T) where T = T
f (generic function with 1 method)

julia> g(::Type{T}) where T = T
g (generic function with 1 method)

julia> f(Int)
DataType

julia> g(Int)
Int64
```

So there's no need to carry that wrapper into the function - `F` is already the type you're looking for, as otherwise you'd get

```
julia> NoInverse(Type{Int})
NoInverse{DataType}(Type{Int64})
```

which surely is not what you want!

Reply:
@Seelengrab , your changes result in

```julia
julia> InverseFunctions.FunctionWithInverse(Complex,Real)
ERROR: MethodError: Cannot `convert` an object of type 
```

With the current state of this PR, we get

```julia
julia> InverseFunctions.FunctionWithInverse(Complex,Real)
(::InverseFunctions.FunctionWithInverse{Type{Complex}, Type{Real}}) (generic function with 1 method)
```

which is what we want.

Reply:
I the end, we want `inverse(some_function_or_type)` and `setinverse(some_function_or_type, other_function_or_type)` to yield properly typed callable function objects.

Reply:
Right - I missed adding `::Type{F}` in the struct definition as well (which would prevent non-type functions, so that's not an option). It's a matter of taste, I guess. With the current state you'll need to add the `Type` unpacking to the signature of `foo` as well, if you want to use that directly, or only go through the field to retrieve the type/function object.

I usually prefer the take-it-from-the-type-parameter approach to save on one field access mentally, but either is fine for performance/optimization.

Reply:
Then again, it's not like you can use `typeof(some_function)` directly to call anything anyway, without already knowing what you're going to call it with - so the current approach should be good :+1:

Reply:
Then again, it's not like you can use `typeof(some_func)` directly to call anything anyway, without knowing what you're going to call it with - so this should be fine either way.

Reply:
> you'll need to add the Type unpacking to the signature of foo as well

@Seelengrab, with unpacking, do you mean

```julia
julia> (;f, invf) = setinverse(Complex, Real)
(::InverseFunctions.FunctionWithInverse{Type{Complex}, Type{Real}}) (generic function with 1 method)

julia> f, invf
(Complex, Real)
```

?

Reply:
Here's an example of what I mean:

```
julia> struct Obj{T}
           o::T
       end

julia> f(::Obj{T}) where T = T(1)
f (generic function with 1 method)

julia> g(::Obj{Type{T}}) where T = T(1)
g (generic function with 1 method)

julia> h(o::Obj) = o.o(1)
h (generic function with 1 method)

julia> f(Obj{Type{Int}}(Int))
ERROR: MethodError: no constructors have been defined for Type{Int64}
Stacktrace:
 [1] f(::Obj{Type{Int64}})
   @ Main ./REPL[1]:1
 [2] top-level scope
   @ REPL[4]:1

julia> g(Obj{Type{Int}}(Int))
1

julia> h(Obj{Type{Int}}(Int))
1
```

You can't really use the `Type{Int}` object directly to construct instances of `Int`; you need the actual `Int` object, which is only stored in the field of your object. The two options to do that are either unpacking the type parameter (bad, because that means you repeat code between `Type` and non-`Type`s who don't need that unpacking) or to always go through the field access (good, because it keeps your code shorter and free from duplication).

Reply:
Actually, come to think of it - what about something like this:

```
struct NoInverse{F, InvF} where {F, InvF}
    NoInverse(f::F, invF::IF) where {F, IF} = new{f, invF}()
end
function(::NoInverse{F, InvF} where {F, InvF} = F
inverse(::NoInverse{F, InvF} where {F, InvF} = InvF
```

, removing the need for `Type` style dispatch to disambiguate types & functions entirely? After all, you can place types in type parameters directly :) 

The downside of course is that you're now forced to carry the type parameters/functions around, but other than that (barring any type instabilities) it should behav the same & compile away just as well.

Reply:
> you're now forced to carry the type parameters/functions around, [...] so I'll just leave this for you all to decide 

Hm, I think I prefer it the way it is now. :-)

Reply:
The approach in https://github.com/JuliaMath/InverseFunctions.jl/pull/32#discussion_r1249593719 does not support callable structs anymore, would it? In any case, I think the current approach is fine for now.